File: dagger-codegen-maven-plugin/src/main/java/org/chelonix/dagger/codegen/introspection/CodegenVisitor.java
package org.chelonix.dagger.codegen.introspection;

import com.squareup.javapoet.*;

import javax.lang.model.element.Modifier;
import java.io.*;
import java.nio.charset.Charset;
import java.nio.file.Path;

import static org.apache.commons.lang3.StringUtils.capitalize;

public class CodegenVisitor implements SchemaVisitor {

    private final ScalarVisitor scalarVisitor;
    private final InputVisitor inputVisitor;
    private final EnumVisitor enumVisitor;
    private final ObjectVisitor objectVisitor;

    private static MethodSpec getter(String var, TypeName type) {
        return MethodSpec.methodBuilder("get" + capitalize(var))
                .addModifiers(Modifier.PUBLIC)
                .returns(type)
                .addStatement("return this.$L", var)
                .build();
    }

    public CodegenVisitor(Schema schema, Path targetDirectory, Charset encoding) {
        this.scalarVisitor = new ScalarVisitor(schema, targetDirectory, encoding);
        this.inputVisitor = new InputVisitor(schema, targetDirectory, encoding);
        this.enumVisitor = new EnumVisitor(schema, targetDirectory, encoding);
        this.objectVisitor = new ObjectVisitor(schema, targetDirectory, encoding);
    }

    @Override
    public void visitScalar(Type type) {
        try {
            scalarVisitor.visit(type);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void visitObject(Type type) {
        try {
            objectVisitor.visit(type);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }



    @Override
    public void visitInput(Type type) {
        try {
            inputVisitor.visit(type);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void visitEnum(Type type) {
        try {
            enumVisitor.visit(type);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: sdk/php/src/Codegen/Introspection/CodegenVisitor.php
<?php

namespace Dagger\Codegen\Introspection;

use GraphQL\Type\Definition\EnumType;
use GraphQL\Type\Definition\InputObjectType;
use GraphQL\Type\Definition\ObjectType;
use GraphQL\Type\Definition\ScalarType;
use GraphQL\Type\Schema;

class CodegenVisitor implements SchemaVisitor
{
    private ScalarVisitor $scalarVisitor;
    private InputVisitor $inputVisitor;
    private EnumVisitor $enumVisitor;
    private ObjectVisitor $objectVisitor;

    public function __construct(
        Schema $schema,
        string $targetDirectory
    ) {
        $this->scalarVisitor = new ScalarVisitor($schema, $targetDirectory);
        $this->inputVisitor = new InputVisitor($schema, $targetDirectory);
        $this->enumVisitor = new EnumVisitor($schema, $targetDirectory);
        $this->objectVisitor = new ObjectVisitor($schema, $targetDirectory);
    }

    public function visitScalar(ScalarType $type): void
    {
        $this->scalarVisitor->visit($type);
    }

    public function visitObject(ObjectType $type): void
    {
        $this->objectVisitor->visit($type);
    }

    public function visitInput(InputObjectType $type): void
    {
        $this->inputVisitor->visit($type);
    }

    public function visitEnum(EnumType $type): void
    {
        $this->enumVisitor->visit($type);
    }
}
